test: add E2E tests for home page, header and footer components

anushkaaaaaaaa opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow our contribution guidelines
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

### **Description:**

I have added E2E tests using Cypress for the homepage, header and footer components to ensure proper functionality and rendering. These tests cover key aspects such as:

- Header: Verifying the presence of essential navigation links, the Async API logo, the star on Github button.
- Footer: Ensuring proper rendering of links and text visibility.
- Home Page: Verifying critical user workflows on the Home Page such as interaction with various links, buttons and logos. 

### Changes:
- Created test cases for the homepage header to check for correct rendering of links to pages present in the Header such as Roadmap, Case Studies and Blog. 
- Added test cases for dropdowns present under Docs, Tools and Community pages.
- Added test cases for the footer to validate that links are rendered correctly.
- Added test cases for the homepage to verify user interaction with various components present.
- Also added test cases to verify links that redirect to different sites from the Home Page present on the Header.

### **Related Issues**-:

- #3671 

<!-- If you refer to a particular issue, provide its number, otherwise, remove this section.
For example, `Resolves #123`, `Fixes #43`, or `See also #33`. The `See also #33` option will not automatically close the issue after the PR merge. -->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- Tests
  - Expanded E2E coverage for the homepage with data-driven navigation to Blog, Docs, Case Studies, Tools, Community, and Roadmap plus many UI/content verifications (header/logo, GitHub star, Read the Docs, feedback link, cards and links).
  - Added header dropdown hover tests for Docs, Tools, and Community.
  - Added footer link validation for initiative, social, news, and Netlify links.
- Chores
  - Added reusable page objects, base helpers, and fixtures to standardize test data and navigation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

anushkaaaaaaaa committed:
feat: add cypress for e2e testing

anushkaaaaaaaa committed:
feat: add cypress for e2e testing

anushkaaaaaaaa committed:
Merge branch 'setup' of https://github.com/anushkaaaaaaaa/website into setup

anushkaaaaaaaa committed:
feat: demo cypress test

anushkaaaaaaaa committed:
Merge branch 'asyncapi:master' into setup

anushkaaaaaaaa committed:
feat: header e2e tests

anushkaaaaaaaa committed:
feat: footer e2e tests

anushkaaaaaaaa committed:
feat: add e2e tests for homepage (and other regarding pages)

anushkaaaaaaaa committed:
Merge branch 'asyncapi:master' into tests

anushkaaaaaaaa committed:
test: add test for docs button

anushkaaaaaaaa committed:
Merge branch 'tests' of https://github.com/anushkaaaaaaaa/website into tests

anushkaaaaaaaa commented:
recording of e2e tests for homepage, header and footer

https://github.com/user-attachments/assets/83a41fba-ace0-4790-8661-24ff99007b35

anushkaaaaaaaa committed:
fix: minor e2e tests fixes

sambhavgupta0705 commented on cypress/header.cy.js:
```diff
@@ -0,0 +1,21 @@
+import Header from './pages/Header';
+
+describe('Footer Links Validation', () => {
```
@anushkaaaaaaaa

anushkaaaaaaaa replied:
> @anushkaaaaaaaa 
> 

Have made the changes.

anushkaaaaaaaa committed:
ci: add e2e tests workflow

anushkaaaaaaaa committed:
Merge remote-tracking branch 'origin/master' into tests

anushkaaaaaaaa committed:
fix: e2e tests fix

anushkaaaaaaaa committed:
fix: minor fixes

anushkaaaaaaaa committed:
ci: minor fix

anushkaaaaaaaa committed:
fix: minor fix

anushkaaaaaaaa committed:
fix: another fix

anushkaaaaaaaa committed:
ci: another minor fix

anushkaaaaaaaa committed:
fix: final fix

akshatnema commented on cypress/pages/DocsPage.js:
Don't you think 🤔, there can be a simplified way of doing this for each page, instead of creating shallow classes for each page?

Think of making a base class for this and then creating these classes inheritng the base class.

anushkaaaaaaaa replied:
I'll be writing a lot more methods(similar to homepage) in each of these classes and hence i've created them separately.

akshatnema commented on cypress/pages/Footer.js:
```diff
@@ -0,0 +1,44 @@
+      { href: 'https://www.twitch.tv/asyncapi', text: 'Twitch' },
+    ];
+  }
+
```
Both components & test files should take centralised data for these assertions. There shouldn't be a case where if code files are changed for this static data, there is a separate overhead to change this data in test files as well.

anushkaaaaaaaa replied:
Thanks for the suggestion That’s a great point I’ll update the code to centralize the social links so both the component and test files use the same source

anushkaaaaaaaa committed:
Merge remote-tracking branch 'origin/master' into tests

anushkaaaaaaaa committed:
fix: failing test fix

anushkaaaaaaaa committed:
test: update tests to get rid of hardcoded links

anushkaaaaaaaa committed:
fix: replace direct Cypress commands with homePage methods

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
Merge branch 'tests' of https://github.com/anushkaaaaaaaa/website into tests

anushkaaaaaaaa committed:
ci: minor fix

anushkaaaaaaaa committed:
fix: minor fix

anushkaaaaaaaa committed:
ci: minor fix

anushkaaaaaaaa committed:
ci: update dependencies

sambhavgupta0705 committed:
Merge branch 'master' into tests

TRohit20 committed:
Merge branch 'master' into tests

sambhavgupta0705 committed:
Merge branch 'master' into tests

sambhavgupta0705 commented:
@anushkaaaaaaaa why we have added casetudies and blog page in this 
like for now we should only have for home and footer I think

anushkaaaaaaaa commented:
> @anushkaaaaaaaa why we have added casetudies and blog page in this 
> like for now we should only have for home and footer I think

The test will fail for visiting case studies page and blog page so I have added those pages( just ) so that it passes as header contains blog page link and case studies page link

anushkaaaaaaaa committed:
fix: small change

anushkaaaaaaaa committed:
test: remove files from tsconfig

anushkaaaaaaaa committed:
Merge branch 'master' into tests

sambhavgupta0705 committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
Merge branch 'master' of https://github.com/anushkaaaaaaaa/website into tests

anushkaaaaaaaa committed:
Merge branch 'tests' of https://github.com/anushkaaaaaaaa/website into tests

anushkaaaaaaaa committed:
Merge branch 'master' into tests

TRohit20 committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
Merge branch 'master' into tests

sambhavgupta0705 committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
feat: implement base classes for each test file

anushkaaaaaaaa committed:
fix: small change

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
fix: small change

sambhavgupta0705 commented on cypress/pages/CommunityPage.js:
```diff
@@ -0,0 +1,6 @@
+class CommunityPage {
+  verifyPageLoaded() {
+    cy.contains('h1','Home of #CommunityOps')
```
what does communityOps do here?

anushkaaaaaaaa replied:
i added it to check if it is visiting the correct page or not; forgot to add the assertion that time, have added that now.

sambhavgupta0705 reviewed (commented):
can we check for all the clrf checks of new line

anushkaaaaaaaa commented:
> can we check for all the clrf checks of new line

sure, will do

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
fix: changes

anushkaaaaaaaa committed:
Merge branch 'tests' of https://github.com/anushkaaaaaaaa/website into tests

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa commented:
> can we check for all the clrf checks of new line

i reviewed the cypress tests and i dont thjnk there are any CRLF or multi-line texts issues so i dont think any changes are required

sambhavgupta0705 commented on cypress/fixtures/footerPageData.json:
```diff
@@ -0,0 +1,24 @@
+  "newsLinks": [
+    { "href": "mailto:press@asyncapi.io", "text": "Email Us" }
+  ],
+  "netlifyLink": "https://netlify.com"
```
why have we added this?

sambhavgupta0705 replied:
like it should be the exact page link imo

anushkaaaaaaaa replied:
okay will make the changes

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
Merge branch 'master' into tests

sambhavgupta0705 committed:
Merge branch 'master' into tests

sambhavgupta0705 reviewed (dismissed):
LGTM!!

sambhavgupta0705 commented:
@coderabbitai please review

akshatnema commented:
@sambhavgupta0705 @anushkaaaaaaaa I see a lot of raw data present in test files or in JSON files, that are directly rendered in the website. Can't we do something to directly take these fixture data from source? Like the website and tests should use same data and in this way, whenever content will be changed, we don't need to change it at 2 places.

sambhavgupta0705 commented:
> @sambhavgupta0705 @anushkaaaaaaaa I see a lot of raw data present in test files or in JSON files, that are directly rendered in the website. Can't we do something to directly take these fixture data from source? Like the website and tests should use same data and in this way, whenever content will be changed, we don't need to change it at 2 places.

@akshatnema we can try having it directly from source by importing the files
Can have a fix around it

sambhavgupta0705 committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
feat: add centralized footer data and update footer validation tests

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
small change in footer.cy.js

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
refactor: simplify visit method in BasePage class

anushkaaaaaaaa committed:
fix: correct import path for footer data in footer tests

anushkaaaaaaaa committed:
refactor: fix lint issues in ToolsCard and ToolsDashboard components

anushkaaaaaaaa committed:
refactor: revert changes in ToolsCard and ToolsDashboard components

anushkaaaaaaaa committed:
small change

anushkaaaaaaaa committed:
Merge branch 'master' into tests

anshgoyalevil committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
small change

anushkaaaaaaaa committed:
fix(footer): update email contact in news links

sambhavgupta0705 committed:
Merge branch 'master' into tests

sambhavgupta0705 committed:
Merge branch 'master' into tests

anushkaaaaaaaa committed:
Merge branch 'master' into tests
